### ISSUE TITLE
Pasted link disappears

### ISSUE THREAD

**Issue by nmielnik** [yabwe/medium-editor#563]
Exposing an issue found by @websirnik in #36 

> I'm testing 4.5.2. With the following options:
> ```js
paste: {
   cleanPastedHTML: true,
   forcePlainText: false
   }
```
> When I paste `<a href="https://google.com">Google</a>`, I expect it would render as a hyperlink [Google](https://google.com), but it renders as a plain-text `Google` (not link). 

> Can you check this behaviour. I'm testing on the` /demo/clean-paste.html` that's included in the repo.

**Comment by nmielnik** [comment]
Also from @websirnik :

> I'm testing on Safari/ChromeCanary. I must say after further testing I got some content pasted correctly. But if you take for instance this [link](http://techcrunch.com/2015/04/15/sources-yahoo-in-talks-to-buy-foursquare/), and copy-paste the 1st paragraph from the article, links does not preserve.

**Comment by nmielnik** [comment]
@websirnik can you pull down the branch I made `paste-fixes` and see if this fixes your issue?  I made some fixes that might help with this.

**Comment by websirnik** [comment]
Testing `paste-fixes` branch. Safari - no results. Chrome - only first links renders, the rest not.
![screen shot 2015-04-16 at 14 44 44](https://cloud.githubusercontent.com/assets/282871/7182142/6310ba7e-e447-11e4-9526-c8c29c5a7e23.png)

**Comment by nmielnik** [comment]
Interesting, I'll try to investigate further.

Once thing to look at in the meantime is to attach to the `paste` event, and inspect what is inside of `event.clipboardData.getData('text/html')` and `event.clipboardData.getData('text/plain')`.

If you can look at that, it might help see the inconsistencies here and lead to a fix.

**Comment by websirnik** [comment]
Hope that helps. Tested on the same article.
Chrome:
![screen shot 2015-04-16 at 15 14 17](https://cloud.githubusercontent.com/assets/282871/7182752/4e6ec4ea-e44b-11e4-966d-df8977722853.png)
Safari:
![screen shot 2015-04-16 at 15 15 15](https://cloud.githubusercontent.com/assets/282871/7182769/759d9b90-e44b-11e4-9c04-38966c969852.png)

**Comment by nmielnik** [comment]
@websirnik so I was able to address some of the issues that you called out, but it looks like things are going to depend on just how you copy and paste.

When I copy and paste the text from the page you sent, I was able to get all the text with the links preserved in chrome (using my branch).

Based on the screen shots you sent, it seems like there are cases where Safari will remove all the html from the clipboard, in which case we would be unable to preserve any html formatting (including links) using our current approach of handling pastes (via inspection of the clipboard).

I've definitely been able to make some improvements, so please try to iterate beyond this, and keep adding more tests as you find more issues, and maybe we can get even closer to a nice paste solution.

**Comment by nmielnik** [comment]
Just to clarify, I was able to pate this into Chrome, Safari, and Firefox on the clean-paste.html demo page, and got the expected results.

```
<meta charset=\'utf-8\'><a href="http://www.yahoo.com/" style="box-sizing: border-box; color: rgb(8, 158, 0); text-decoration: none; outline: 0px; font-family: \'Open Sans\', HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 16px; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: 26px; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 1; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(249, 249, 249);">Yahoo</a><span style="color: rgb(62, 67, 62); font-family: \'Open Sans\', HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 16px; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: 26px; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 1; word-spacing: 0px; -webkit-text-stroke-width: 0px; display: inline !important; float: none; background-color: rgb(249, 249, 249);"><span class="Apple-converted-space"> </span>has been busy rebuilding its business around<span class="Apple-converted-space"> </span></span><a href="http://www.fastcompany.com/3044281/marissa-mayer" style="box-sizing: border-box; color: rgb(8, 158, 0); text-decoration: none; outline: 0px; font-family: \'Open Sans\', HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 16px; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: 26px; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 1; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(249, 249, 249);">mobile</a><span style="color: rgb(62, 67, 62); font-family: \'Open Sans\', HelveticaNeue, Helvetica, Arial, sans-serif; font-size: 16px; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: 26px; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 1; word-spacing: 0px; -webkit-text-stroke-width: 0px; display: inline !important; float: none; background-color: rgb(249, 249, 249);"> under CEO Marissa Mayer, and soon it could make one of its biggest bets yet on the platform.</span>
```

**Comment by websirnik** [comment]
Here is more debug information(Mac OS X 10.10.3):

:+1: I got a successful paste when the ***copy*** was initiated from Chrome/Firefox 

:-1: I got the same results as here: https://github.com/daviferreira/medium-editor/issues/563#issuecomment-93745500 when the ***copy*** was initiated from Safari.

**Comment by nmielnik** [comment]
Yeah, that makes sense based on the screenshot you showed (the `clipboardData.getData('text/html')` call returned nothing, and the `clipboardData.getData('text/plain')` returned plain-text).  It seems as though copies from Safari don't allow us to access the HTML, thus we can't preserve the anchors (or any HTML for that matter).

**Comment by websirnik** [comment]
Just checked copy-paste on [medium.com](medium.com),  Safari does preserve links there for the same content.

**Comment by nmielnik** [comment]
Yeah, they're implementation will likely be different than ours in that they don't inspect the clipboard on paste.  My guess is they're likely allowing the paste to just fall through, and then attempting to do some cleanup after the fact.  It's a limitation of using the clipboard, so we'd need to redo how we handle pasting to account for this.

**Comment by nmielnik** [comment]
This should now be fixed for keyboard paste by #973.  Fixing for right-click/context-menu paste is a whole other monster.
